sdks/python/apache_beam/transforms/stats.py

```diff
@@ -0,0 +1,236 @@
+# limitations under the License.
+#
+
+"""This module has all statistic related transforms."""
+
+from __future__ import absolute_import
+from __future__ import division
+
+import heapq
+import math
+import sys
+from builtins import round
+
+import mmh3
+
+from apache_beam import coders
+from apache_beam import typehints
+from apache_beam.transforms.core import *
+from apache_beam.transforms.ptransform import PTransform
+
+__all__ = [
+    'ApproximateUnique',
+]
+
+# Type variables
+T = typehints.TypeVariable('T')
+K = typehints.TypeVariable('K')
+V = typehints.TypeVariable('V')
+
+
+class ApproximateUnique(object):
+  """
+  Hashes input elements and uses those to extrapolate the size of the entire
+  set of hash values by assuming the rest of the hash values are as densely
+  distributed as the sample space.
+  Args:
+  **kwargs: Accepts a single named argument "size" or "error".
+  size: an int not smaller than 16, which we would use to estimate number of
+  unique values.
+  error: max estimation error, which is a float between 0.01 and 0.50. If
+  error is given, size will be calculated from error with
+   _get_sample_size_from_est_error function.
+
+ """
+
+  _NO_VALUE_ERR_MSG = 'Either size or error should be set. Received {}.'
+  _MULTI_VALUE_ERR_MSG = 'Either size or error should be set. ' \
+                         'Received {size = %s, error = %s}.'
+  _INPUT_SIZE_ERR_MSG = 'ApproximateUnique needs a size >= 16 for an error ' \
+                        '<= 0.50. In general, the estimation error is about ' \
+                        '2 / sqrt(sample_size). Received {size = %s}.'
+  _INPUT_ERROR_ERR_MSG = 'ApproximateUnique needs an estimation error ' \
+                         'between 0.01 and 0.50. Received {error = %s}.'
+
+  @staticmethod
+  def get_input_params(size=None, error=None):
+    """
+    :return: sample size
+
+    Check if input params are valid and return sample size.
+    """
+    if None not in (size, error):
+      raise ValueError(ApproximateUnique._MULTI_VALUE_ERR_MSG % (size, error))
+    elif size is None and error is None:
+      raise ValueError(ApproximateUnique._NO_VALUE_ERR_MSG)
+    elif size is not None:
+      if not isinstance(size, int) or size < 16:
+        raise ValueError(ApproximateUnique._INPUT_SIZE_ERR_MSG % (size))
+      else:
+        return size
+    else:
+      if error < 0.01 or error > 0.5:
+        raise ValueError(ApproximateUnique._INPUT_ERROR_ERR_MSG % (error))
+      else:
+        return ApproximateUnique._get_sample_size_from_est_error(error)
+
+  @staticmethod
+  def _get_sample_size_from_est_error(est_err):
+    """
+    :return: sample size
+
+    Calculate sample size from estimation error
+    """
+    #math.ceil in python2.7 returns a float, while it returns an int in python3.
+    return int(math.ceil(4.0 / math.pow(est_err, 2.0)))
+
+  @typehints.with_input_types(T)
+  @typehints.with_output_types(T)
```

Comment:
I think the output type here should be a numeric type, not the same input type `T`. Same below.